providers/providers.go

```diff
@@ -32,6 +41,16 @@ type Provider interface {
 }
 
 func NewProvider(providerConfig options.Provider) (Provider, error) {
+	if providerConfig.OIDCConfig.EmailClaim == "" {
+		providerConfig.OIDCConfig.EmailClaim = DefaultEmailClaim
+	}
+	if providerConfig.OIDCConfig.GroupsClaim == "" {
+		providerConfig.OIDCConfig.GroupsClaim = DefaultGroupsClaim
+	}
+	if len(providerConfig.OIDCConfig.AudienceClaims) == 0 {
+		providerConfig.OIDCConfig.AudienceClaims = DefaultAudienceClaims
+	}
+
```

Comment:
Unfortunately, this isn't a good solution. We do already have logic in place for the legacy and alpha config to properly set the OIDC default values. You seem to have found a bug with the existing logic. Instead of adding this additional code we need to check the logic inside `options.go` to see what is going on.

Furthermore, I know we unfortunately have quite a bit of OIDC related code inside `providers.go` which shouldn't be the case. This file should be as generic and uncoupled of any provider as possible.

Reply:
> Instead of adding this additional code we need to check the logic inside options.go to see what is going on.

`legacyProviderFlagSet()` has defaults. But `loadAlphaOptions()` overwrites `opts.Providers` without validation.
Are you saying that `pkg.validation.providers.validateProvider()` is responsible for the validation?

> Furthermore, I know we unfortunately have quite a bit of OIDC related code inside providers.go which shouldn't be the case. This file should be as generic and uncoupled of any provider as possible.

If the documentation said it was the default, this could be verified when interpreting options, but I think `New()` is responsible for verification by default.

The configuration of oauth2-proxy is huge, and I know that options are basically controlled by flags and files, and I agree that viper and pflag are good packages in this regard.

However, a lot of documentation for goper suggests setting defaults via the `New()` when a default value is used for a struct, and I didn't think that was odd considering that many packages use the following design pattern.

```go
type OptionFunc func(*Struct)

func WithValueOpt(value string) OptionFunc {
	if value == "" {
		panic("not be empty")
	}
	return  func(s *Struct) {
		s.Value = value
	}
}

func New(opts ...OptionFunc) {
	// ...
```

Reply:
I think you misunderstood me. What I'm saying is that not all providers that oauth2-proxy offers are OIDC providers. Therefore the defaults for OIDC should not be set in the generic NewProvider but rather in the respective methods dedicated to initialize the defaults for the OIDC config. This should already be the case but as I said it seems like you found a bug in the alpha config implementation.

Reply:
Oh, I got it. When I was working on it, I had considered initializing inside `providers.NewOIDCProvider()` and calling it internally in functions like `providers.NewKeycloakOIDCProvider()`.

But, `NewProvider()` calls `newProviderDataFromConfig()` before `NewOIDCProvider()`. And  `newProviderDataFromConfig()` uses `OIDCConfig.xxx`.

If want to initialize something in `NewOIDCProvider()`. We need to refactor `NewProvider`.